prediction_market_agent/agents/known_outcome_agent/known_outcome_agent.py

```diff
@@ -0,0 +1,165 @@
+
+where <REASONING> is a free text field containing your reasoning, and <RESULT>
+is a multiple-choice field containing only one of 'YES' (if the answer to the
+question is yes), 'NO' (if the answer to the question is no), or 'UNKNOWN' if
+you are unable to answer the question with a reasonable degree of certainty from
+the web-scraped information. Your answer should only contain this json string,
+and nothing else.
+
+If the question is of the format: "Will X happen by Y?" then the result should
+be as follows:
+- If X has already happened, the result is 'YES'.
+- If not-X has already happened, the result is 'NO'.
+- If X has been announced to happen after Y, result 'NO'.
+- Otherwise, the result is 'UNKNOWN'.
+
+If the question is of the format: "Will X happen on Y?"
+- If something has happened that necessarily prevents X from happening on Y, the result is 'NO'.
+- Otherwise, the result is 'UNKNOWN'.
+"""
+
+
+def completion_str_to_json(
+    completion: str,
+) -> dict[str, t.Any]:  # noqa: F811 # Allow t.Any in the return type
+    """
+    Cleans completion JSON in form of a string:
+
+    ```json
+    {
+        ...
+    }
+    ```
+
+    into just { ... }
+    ```
+    """
+    start_index = completion.find("{")
+    end_index = completion.rfind("}")
+    completion = completion[start_index : end_index + 1]
+    return json.loads(completion)  # type: ignore # Allow t.Any in the return type
+
+
+def get_known_outcome(model: str, question: str, max_tries: int) -> Answer:
+    """
+    In a loop, perform web search and scrape to find if the answer to the
+    question is known. Break if the answer is found, or after a certain number
+    of tries, and no definite answer is found, return an 'unknown' answer.
+    """
+    tries = 0
+    date_str = datetime.now().strftime("%d %B %Y")
+    previous_urls = []
+    llm = ChatOpenAI(model="gpt-4-1106-preview", temperature=0.4)
+    while tries < max_tries:
+        search_prompt = ChatPromptTemplate.from_template(
+            template=GENERATE_SEARCH_QUERY_PROMPT
+        ).format_messages(date_str=date_str, question=question)
+        search_query = str(llm.invoke(search_prompt).content).strip('"')
+        search_results = web_search(query=search_query, max_results=5)
+        if not search_results:
+            raise ValueError("No search results found.")
+
+        for result in search_results:
+            if result.url in previous_urls:
+                continue
+            previous_urls.append(result.url)
+
+            scraped_content = web_scrape(url=result.url)
+
+            prompt = ChatPromptTemplate.from_template(
+                template=ANSWER_FROM_WEBSCRAPE_PROMPT
+            ).format_messages(
+                date_str=date_str,
+                question=question,
+                scraped_content=scraped_content,
+            )
+            answer = str(llm.invoke(prompt).content)
+            parsed_answer = Answer.model_validate(completion_str_to_json(answer))
```

Comment:
Isn't it a bit risky to assume the LLM will answer using the correct format? 
I would expect some try-catch block because of this, but I might be wrong.

Reply:
I'd say it's fine it depends on what comes after this.

Is this needed (if so let's error out and do it more robust in another PR) or can we skip it and continue ?

Reply:
gpt4 is very reliable at producing json output!